doc/developer_guide/psy_data.rst

```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+The ``FieldDouble`` and ``FieldVectorDouble`` related functions are explicitly
+coded in the library, the rest is taken from the base class:
+
+.. code-block:: jinja
+
+    procedure :: DeclareFieldDouble,  ProvideFieldDouble
+    procedure :: DeclareFieldVectorDouble,  ProvideFieldVectorDouble
+
+    {# Collect the various procedures for the same generic interface #}
+    {# ------------------------------------------------------------- #}
+    {% set all_declares=[] %}
+    {% set all_provides=[] %}
+    {% for name, type, bits in ALL_TYPES %}
+      {{ all_declares.append("DeclareScalar"~name) or "" -}}
+      {{ all_provides.append("ProvideScalar"~name) or "" -}}
+      {% for dim in ALL_DIMS %}
+        {{ all_declares.append("DeclareArray"~dim~"d"~name) or "" -}}
+        {{ all_provides.append("ProvideArray"~dim~"d"~name) or "" -}}
+      {% endfor %}
+    {% endfor %}
+
+    {% set indent="            " %}
+    generic, public :: PreDeclareVariable => &
+        DeclareFieldDouble, &
+        DeclareFieldVectorDouble, &
+        {{all_declares|join(", &\n"+indent) }}
+
+    generic, public :: ProvideVariable => &
+        ProvideFieldDouble,       &
+        ProvideFieldVectorDouble, &
+        {{all_provides|join(", &\n"+indent) }}
+
+.. note:
+    Using ``or ""`` in the jinja statements avoids that ``None`` is
+    added to the files (which would be the output of e.g. the append
+    instruction). The ``-`` before the closing ``}}`` prevents this
+    line from creating any white-spaces. As a result of this the
+    processed file will not have unusual empty lines or indentation.
+
+
+Static Functions in the Base Class
+++++++++++++++++++++++++++++++++++
+The base class provides the four static functions defined in the API (see
+:ref:`psy_data_api`). The subroutines ``PREFIX_PSyDataInit()`` and
+``PREFIX_PSyDataShutdown()`` are empty, and ``PREFIX_PSyDataStart()`` and
+``PREFIX_PSyDataStop()`` function just change the module variable
+``is_enabled`` (which can then be used by a wrapper library to enable
+or disable its functionality).
+
+If you don't need specific functionality for these functions, you
+can just declare them in your wrapper library::
+
+    module MyProfileWrapperLibrary
+      use psydata_base_mod, only : profile_PSyDataInit, profilePSyDataShutdown, &
+                                   profile_PSyDataStart, profile_PSyDataStop
+
+If you need to partially change the behaviour of these functions,
+you have to rename them in your ``use`` statement, and then call the renamed
+base class function like this::
+
+    module MyProfileWrapperLibrary
+    ...
+    contains
+
+    subroutine profile_PSyDataStart()
+      use psydata_base_mod, only : base_PSyDataStart => profile_PSyDataStart
+      ! Do something
+      call basePSyDataStart()
+      ! Do something else
+    end subroutine profile_PSyDataStart
+
+.. note:
+    The ``PSyDataBase`` class will use the prefix that you have specified
```

Comment:
I've extended this a bit, hope it's better. The basic idea is that if you need to provide your own version of one of the static functions (but still need to use the functionality of the original class), you have to rename the original function somehow - either you create it using a different name, or you rename it when importing ('use'). I've opted for renaming, but thought it would be useful that you could also just change all prefixes in the base class using jinja.

Reply:
Oh yes, I get it now :-)